File: common/src/main/java/com/hartwig/actin/serialization/MolecularRecordJson.java
package com.hartwig.actin.serialization;

import static com.hartwig.actin.serialization.JsonFunctions.array;
import static com.hartwig.actin.serialization.JsonFunctions.bool;
import static com.hartwig.actin.serialization.JsonFunctions.object;
import static com.hartwig.actin.serialization.JsonFunctions.string;

import java.io.File;
import java.io.IOException;
import java.lang.reflect.Type;
import java.nio.file.Files;
import java.util.List;

import com.google.common.collect.Lists;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonArray;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;
import com.hartwig.actin.datamodel.molecular.EvidenceDirection;
import com.hartwig.actin.datamodel.molecular.EvidenceLevel;
import com.hartwig.actin.datamodel.molecular.GenomicTreatmentEvidence;
import com.hartwig.actin.datamodel.molecular.ImmutableGenomicTreatmentEvidence;
import com.hartwig.actin.datamodel.molecular.ImmutableMolecularRecord;
import com.hartwig.actin.datamodel.molecular.MolecularRecord;

import org.jetbrains.annotations.NotNull;

public final class MolecularRecordJson {

    private MolecularRecordJson() {
    }

    @NotNull
    public static MolecularRecord read(@NotNull String molecularJson) throws IOException {
        Gson gson = new GsonBuilder().registerTypeAdapter(MolecularRecord.class, new MolecularRecordCreator()).create();
        String json = Files.readString(new File(molecularJson).toPath());
        return gson.fromJson(json, MolecularRecord.class);
    }

    private static class MolecularRecordCreator implements JsonDeserializer<MolecularRecord> {

        @Override
        public MolecularRecord deserialize(@NotNull JsonElement jsonElement, @NotNull Type type,
                @NotNull JsonDeserializationContext jsonDeserializationContext) throws JsonParseException {
            JsonObject record = jsonElement.getAsJsonObject();

            JsonObject purple = object(record, "purple");
            return ImmutableMolecularRecord.builder()
                    .sampleId(string(record, "sampleId"))
                    .hasReliableQuality(bool(purple, "hasReliableQuality"))
                    .hasReliablePurity(bool(purple, "hasReliablePurity"))
                    .genomicTreatmentEvidences(toGenomicTreatmentEvidences(array(record, "protect")))
                    .build();
        }

        @NotNull
        private static List<GenomicTreatmentEvidence> toGenomicTreatmentEvidences(@NotNull JsonArray protect) {
            List<GenomicTreatmentEvidence> evidences = Lists.newArrayList();
            for (JsonElement element : protect) {
                JsonObject evidence = element.getAsJsonObject();
                boolean reported = bool(evidence, "reported");
                if (reported) {
                    evidences.add(ImmutableGenomicTreatmentEvidence.builder()
                            .genomicEvent(string(evidence, "genomicEvent"))
                            .treatment(string(evidence, "treatment"))
                            .onLabel(bool(evidence, "onLabel"))
                            .level(EvidenceLevel.valueOf(string(evidence, "level")))
                            .direction(EvidenceDirection.valueOf(string(evidence, "direction")))
                            .build());
                }
            }
            return evidences;
        }
    }
}

File: common/src/main/kotlin/com/hartwig/actin/molecular/serialization/MolecularRecordJson.kt
package com.hartwig.actin.molecular.serialization

import com.google.gson.GsonBuilder
import com.google.gson.reflect.TypeToken
import com.hartwig.actin.molecular.datamodel.MolecularRecord
import com.hartwig.actin.util.Paths
import com.hartwig.actin.util.json.GsonLocalDateAdapter
import com.hartwig.actin.util.json.GsonSerializer
import org.apache.logging.log4j.LogManager
import java.io.BufferedWriter
import java.io.File
import java.io.FileWriter
import java.nio.file.Files
import java.time.LocalDate

object MolecularRecordJson {

    private val LOGGER = LogManager.getLogger(MolecularRecordJson::class.java)
    private const val MOLECULAR_JSON_EXTENSION = ".molecular.json"

    fun write(record: MolecularRecord, directory: String) {
        val path = Paths.forceTrailingFileSeparator(directory)
        val jsonFile = path + record.sampleId + MOLECULAR_JSON_EXTENSION
        LOGGER.info("Writing molecular record to {}", jsonFile)
        val writer = BufferedWriter(FileWriter(jsonFile))
        writer.write(toJson(record))
        writer.close()
    }

    fun read(molecularJson: String): MolecularRecord {
        return fromJson(Files.readString(File(molecularJson).toPath()))
    }

    fun toJson(record: MolecularRecord): String {
        return GsonSerializer.create().toJson(record)
    }

    fun fromJson(json: String): MolecularRecord {
        return GsonBuilder().serializeNulls()
            .enableComplexMapKeySerialization()
            .registerTypeAdapter(object : TypeToken<LocalDate?>() {}.type, GsonLocalDateAdapter())
            .create()
            .fromJson(json, MolecularRecord::class.java)
    }
}
